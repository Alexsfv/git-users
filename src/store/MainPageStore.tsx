import { action, makeObservable, observable, runInAction } from "mobx"
import { GithubResponseType, GithubUserItem } from "../types/types"

export class MainPageStore {
    userValue: string = ''
    users: Array<GithubUserItem> = []
    isLoadedUsers: boolean = false
    errorInputMessage: string = ''

    constructor() {
        makeObservable(this, {
            userValue: observable,
            users: observable,
            isLoadedUsers: observable,
            errorInputMessage: observable,
            getUsersFormHandler: action,
            getUsers: action,
            changeUserValue: action,
            changeError: action,
            setIsLoadedUsers: action,
            setUsers: action,
        })
    }

    getUsersFormHandler: React.FormEventHandler = (event) => {
        event.preventDefault()
        this.getUsers()
    }

    changeUserValue: React.ChangeEventHandler<HTMLInputElement> = (event) => {
        this.userValue = event.target.value
        if (this.errorInputMessage) this.changeError('')
    }

    getUsers = async () => {
        try {
            if (this.userValue.length <= 0) {
                this.changeError('Введите имя пользователя')
            } else {
                const url = `https://api.github.com/search/users?q=${this.userValue}&`
                const response: GithubResponseType = await fetch(url).then(res => res.json())
                const users = response.items
                runInAction(() => {
                    this.setUsers(users)
                    this.setIsLoadedUsers(true)
                })
            }
        } catch (e) {
            console.warn('Get users request error: ', e.message)
        }
    }

    setUsers(data: GithubUserItem[]) { this.users = data }

    setIsLoadedUsers(status: boolean) { this.isLoadedUsers = status }

    changeError(message: string) { this.errorInputMessage = message }
}
