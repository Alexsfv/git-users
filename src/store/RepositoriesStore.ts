import { GithubRepository } from './../types/types';
import { makeObservable, observable, runInAction, action } from 'mobx';

export class RepositoriesStore {
    repos: GithubRepository[] = []
    repos_count: number = 0

    constructor() {
        makeObservable(this, {
            repos: observable,
            repos_count: observable,
            getRepos: action
        })
    }

    getRepos = async (login: string) => {
        try {
            const url = `https://api.github.com/users/${login}/repos`
            const response = await fetch(url)

            if (response.status === 200) {  
                const repos = await response.json()
                runInAction(() => {
                    this.repos = repos
                    this.repos_count = repos.length
                })

            }

        } catch (e) {

        }
    }
}