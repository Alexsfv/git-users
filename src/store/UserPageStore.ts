import { runInAction } from 'mobx';
import { observable } from 'mobx';
import { action } from 'mobx';
import { GithubUserPage } from './../types/types';
import { makeObservable } from 'mobx';
import { getTimeString } from '../types/func/time';

export interface IUserPageStore extends GithubUserPage {
    isLoaded: boolean
    isExists: boolean | null
    errorMessage: string
    getUserData: (login: string) => void
    clearUserPage: () => void
    catchErrorRosponse: (status: number) => void
}

const userPageStore: IUserPageStore = {
    isLoaded: false,
    isExists: null,
    errorMessage: '',
    login: '',
    avatar_url: '',
    name: '',
    location: null,
    email: null,
    public_repos: 0,
    public_gists: 0,
    followers: 0,
    following: 0,
    blog: '',
    bio: null,
    created_at: '',
    updated_at: '',
    async getUserData(login) {
        try {
            const url = `https://api.github.com/users/${login}`
            const response = await fetch(url)
            if (response.status === 200) {
                const userData: GithubUserPage = await response.json()
                runInAction(() => {
                    this.isExists = true
                    this.errorMessage = ''
                    this.login = userData.login
                    this.avatar_url = userData.avatar_url
                    this.name = userData.name
                    this.location = userData.location
                    this.email = userData.email
                    this.public_repos = userData.public_repos
                    this.public_gists = userData.public_gists
                    this.followers = userData.followers
                    this.following = userData.following
                    this.blog = userData.blog
                    this.bio = userData.bio
                    this.created_at = getTimeString(userData.created_at)
                    this.updated_at = getTimeString(userData.updated_at)
                })
            } else {
                this.catchErrorRosponse(response.status)
            }
        } catch (e) {
            console.warn(e.message)
        }
    },
    catchErrorRosponse(status: number) {
            if (status === 404) {
                runInAction(() => {
                    this.isExists = false
                    this.errorMessage = 'Пользователь с таким логином не существует'
                })
            } else if (status === 403) {
                runInAction(() => {
                    this.errorMessage = 'Превышен лимит запросов, попробуйте позже'
                })
            } else {
                runInAction(() => {
                    this.errorMessage = 'Что-то пошло не так'
                })
            }
    },
    clearUserPage() {
        this.isLoaded = false
        this.isExists = false
        this.errorMessage = ''
        this.login = ''
        this.avatar_url = ''
        this.name = ''
        this.location = null
        this.email = null
        this.public_repos = 0
        this.public_gists = 0
        this.followers = 0
        this.following = 0
        this.blog = ''
        this.bio = null
        this.created_at = ''
        this.updated_at = ''
    }
}

const userPageAnnotations = {
    isLoaded: observable,
    isExists: observable,
    errorMessage: observable,
    login: observable,
    avatar_url: observable,
    name: observable,
    location: observable,
    email: observable,
    public_repos: observable,
    public_gists: observable,
    followers: observable,
    following: observable,
    blog: observable,
    bio: observable,
    created_at: observable,
    updated_at: observable,
    getUserData: action.bound,
    clearUserPage: action.bound,
    catchErrorRosponse: action.bound
}


export const UserPageStore = (): IUserPageStore => makeObservable(userPageStore, userPageAnnotations)