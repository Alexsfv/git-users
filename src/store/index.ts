export * from './MainPageStore'
export * from './UserPageStore'
export * from './RepositoriesStore'