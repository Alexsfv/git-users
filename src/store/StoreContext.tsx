import React, { createContext, useContext } from 'react'
import { rootStoreType, StoreContextProps } from '../types/types'
import { MainPageStore, UserPageStore } from './index'
import { RepositoriesStore } from './RepositoriesStore'

const rootStore: rootStoreType = {
    mainPage: new MainPageStore(),
    userPage: UserPageStore(),
    repositories: new RepositoriesStore()
}

const storeContext = createContext<rootStoreType | null>(null)

export const StoreContext: React.FC<StoreContextProps> = ({children}) => {
    return (
        <storeContext.Provider value={rootStore}>
            {children}
        </storeContext.Provider>
    )
}

export const useStore = () => {
    const store = useContext(storeContext)
    if (store === null) {
        throw new Error('You call useContext, but component has not context. Try use context.Provider component')
    }
    return store
}
