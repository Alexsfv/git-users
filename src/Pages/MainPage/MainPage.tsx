import { observer } from 'mobx-react'
import React from 'react'
import InputUser from '../../Components/InputUser/InputUser'
import UserCard from '../../Components/UserCard/UserCard'
import { useStore } from '../../store/StoreContext'



const MainPage: React.FC<{}> = observer(() => {

    const { mainPage } = useStore()

    return (
        <section>
            <div className="container">
                <div className="row d-flex mt-3 justify-content-between">

                    <InputUser />
                    {
                        mainPage.isLoadedUsers
                            ? mainPage.users.map(user => {
                                return (
                                    <UserCard 
                                        key={user.id}
                                        user={user}
                                    />
                                )
                            })
                            : <div className="d-flex w-100">
                                <p 
                                    className="w-100 text-center text-success font-weight-bold"
                                    style={{fontSize: '24px'}}
                                >
                                    Данное приложение осуществляет поиск GitHub акаунтов по логину пользователя
                                </p>
                            </div>
                    }

                </div>
            </div>
        </section>
    )
})

export default MainPage