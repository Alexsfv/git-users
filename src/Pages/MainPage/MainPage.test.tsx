import { render } from 'enzyme'
import React from 'react'
import { BrowserRouter } from 'react-router-dom'
import { rootStore, StoreContext } from '../../store/StoreContext'
import MainPage from './MainPage'

describe("MainPage component", () => {
    it("should render placeholder data", () => {
        const jsx = (
            <StoreContext>
                <MainPage/>
            </StoreContext>
        )
        const component = render(jsx)
        expect(component).toMatchSnapshot()
    })

    it("should render UserCard components", () => {
        const user = {
            avatar_url: "https://google.com",
            id: 55,
            login: 'user-login'
        }
        rootStore.mainPage.setIsLoadedUsers(true)
        rootStore.mainPage.setUsers([user, user])
        const jsx = (
            <BrowserRouter>
                <StoreContext>
                    <MainPage/>
                </StoreContext>
            </BrowserRouter>
        )
        const component = render(jsx)
        expect(component).toMatchSnapshot()
    })
})