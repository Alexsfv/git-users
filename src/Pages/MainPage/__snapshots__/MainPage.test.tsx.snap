// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage component should render UserCard components 1`] = `
<section>
  <div
    class="container"
  >
    <div
      class="row d-flex mt-3 justify-content-between"
    >
      <form
        class="w-100 mb-5"
      >
        <input
          class="form-control form-control-lg"
          placeholder="Введите имя пользователя"
          type="text"
          value=""
        />
        <div
          class=""
        />
        <input
          class="btn w-100 btn-success mt-3"
          type="submit"
          value="Найти"
        />
      </form>
      <div
        class="card mb-5"
        style="width:18rem"
      >
        <img
          alt="userAvatar"
          class="card-img-top"
          src="https://google.com"
        />
        <div
          class="card-body"
        >
          <h5
            class="card-title"
          >
            user-login
          </h5>
          <a
            class="btn btn-primary"
            href="/userPage/user-login"
          >
            Открыть профиль
          </a>
        </div>
      </div>
      <div
        class="card mb-5"
        style="width:18rem"
      >
        <img
          alt="userAvatar"
          class="card-img-top"
          src="https://google.com"
        />
        <div
          class="card-body"
        >
          <h5
            class="card-title"
          >
            user-login
          </h5>
          <a
            class="btn btn-primary"
            href="/userPage/user-login"
          >
            Открыть профиль
          </a>
        </div>
      </div>
    </div>
  </div>
</section>
`;

exports[`MainPage component should render placeholder data 1`] = `
<section>
  <div
    class="container"
  >
    <div
      class="row d-flex mt-3 justify-content-between"
    >
      <form
        class="w-100 mb-5"
      >
        <input
          class="form-control form-control-lg"
          placeholder="Введите имя пользователя"
          type="text"
          value=""
        />
        <div
          class=""
        />
        <input
          class="btn w-100 btn-success mt-3"
          type="submit"
          value="Найти"
        />
      </form>
      <div
        class="d-flex w-100"
      >
        <p
          class="w-100 text-center text-success font-weight-bold"
          style="font-size:24px"
        >
          Данное приложение осуществляет поиск GitHub акаунтов по логину пользователя
        </p>
      </div>
    </div>
  </div>
</section>
`;
